Add a Toggle to the Status Bars w/ a Delay to show only in Combat

ermalsh opened the pull request:
**Duplicate of #7445 due to Orphaned Source branch.** 

Upon many emails and several requests to fix the review suggestions made by @Nightfirecat in #7445 , I came back to clean up those review items in hopes to have this merged into a release for all those people that have asked.

_Duplicate PR content:_
This is to only show the Status Bars only when in Combat, and hides it based on the delay set.
Closes #7436
Toggle Demo: https://gyazo.com/ee277e5f0d4bfad598a7a1049d972a67
Delay Demo: https://gyazo.com/421e2a4a725725f1b7b5e62f96a6a8a3

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -60,4 +61,15 @@ default boolean enableRestorationBars()
+	@ConfigItem(
+		keyName = "hideAfterCombatDelay",
+		name = "Hide after combat delay",
+		description = "Time required to hide status bars after no longer in combat."
```
This should mention that `0` will mean bars are always shown

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -47,10 +59,18 @@
+	@Inject
+	private StatusBarsConfig config;
+
+	@Getter(AccessLevel.PACKAGE)
```
This field isn't accessed outside this class. Why does it have a getter?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -60,4 +61,15 @@ default boolean enableRestorationBars()
+		name = "Hide after combat delay",
+		description = "Time required to hide status bars after no longer in combat."
+	)
+	@Units(Units.SECONDS)
```
Game ticks would be a better unit of measurement since that's the unit of time this change actually operates on. That would also let you drop the `Instant.now()` duration comparisons in favor of using `client.getTickCount()`.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -47,10 +59,18 @@
 	@Override
 	protected void startUp() throws Exception
 	{
-		overlayManager.add(overlay);
```
`startUp()` should be removed if it contains no code.

ermalsh commented:
Should be all set with the recommended changes. 

Requesting additional review from @Nightfirecat

Nightfirecat reviewed (commented):
I've done a little cleanup and added a PVP combat check. Can you confirm this is still working as expected?

ermalsh commented:
> I've done a little cleanup and added a PVP combat check. Can you confirm this is still working as expected?

@Nightfirecat Thank you for doing that, I have confirmed it is still working as expected after a fetch of your changes - see GIF attached.

![statusbars](https://user-images.githubusercontent.com/19981656/96156547-32efbc00-0edf-11eb-92cc-fc06bda31e5e.gif)

Malfuryent commented:
#7730 is still waiting to get merged into the client and would probably take priority. this PR is not valid anymore once the updated statusbars get added into the client. Maybe #7730 could be merged first so this one could be adjusted to the new statusbars?

ermalsh commented:
> #7730 is still waiting to get merged into the client and would probably take priority. this PR is not valid anymore once the updated statusbars get added into the client. Maybe #7730 could be merged first so this one could be adjusted to the new statusbars?

Agreed ^

Nightfirecat commented:
#7730 has been merged, this can be rebased to fix conflicts with it. Can you also verify this works in PVP combat? (LMS is an easy place to test this)

ermalsh commented:
> #7730 has been merged, this can be rebased to fix conflicts with it. Can you also verify this works in PVP combat? (LMS is an easy place to test this)

Rebased and fixed conflicts, confirmed working in PVP via LMS.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -54,14 +56,25 @@ default boolean enableSkillIcon()
 		keyName = "enableRestorationBars",
-		name = "Show restores",
-		description = "Visually shows how much will be restored to your status bar."
+		name = "Show amount of hitpoints and prayer restored",
```
Usually we prefer short names since there is not much room for it in the side panel. When you hover over it it will show the description anyway. This would probably just show `Show amount...`

ermalsh replied:
Agreed! Change made.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsConfig.java:
```diff
@@ -54,14 +56,25 @@ default boolean enableSkillIcon()
 
+	@ConfigItem(
+		keyName = "hideAfterCombatDelay",
+		name = "Hide after combat delay",
```
Maybe `Hide delay`

Adam- commented:
Instead of all of this complexity, you can just store the last combat tick and check it in the overlay.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -64,4 +74,39 @@ StatusBarsConfig provideConfig(ConfigManager configManager)
+		}
+		else
+		{
+			overlayManager.add(overlay);
```
Instead of adding and removing the overlay various places, you can keep the overlay enabled the entire time and `return;` from it early when the combat has timed out.

Nightfirecat commented:
I've cleaned this up a bit and addressed Adam's comment

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -47,21 +59,78 @@
+	{
+		if (StatusBarsConfig.GROUP.equals(event.getGroup()) && event.getKey().equals("hideAfterCombatDelay"))
+		{
+			shouldToggleStatusBars();
```
This must be run on the client thread or will error.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -47,21 +59,78 @@
+		}
+	}
+
+	private void shouldToggleStatusBars()
```
This should be named something which resembles what it is doing like ` checkStatusBars`

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/statusbars/StatusBarsPlugin.java:
```diff
@@ -47,21 +59,78 @@
+			lastCombatActionTickCount = client.getTickCount();
+			barsDisplayed = true;
+		}
+		else if (config.hideAfterCombatDelay() == 0)
```
I think you can just check this above and do an early return if it is 0, since it isn't necessary to check anything else then?
